src/containers/rbtree/rbtree.h

```diff
@@ -0,0 +1,794 @@
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+template <typename K, typename T>
+class rbtree {
+ public:
+  using key_type = K;
+  using mapped_type = T;
+  using value_type = std::pair<const key_type, mapped_type>;
+  using reference = value_type&;
+  using const_reference = const value_type&;
+  using size_type = size_t;
+  using pointer = value_type*;
+  struct rbtree_node {
+    rbtree_node(const value_type& value)
+        : value_(value),
+          right_(nullptr),
+          left_(nullptr),
+          parent_(nullptr),
+          mColor('R') {}
+    value_type value_;
+    rbtree_node* right_;
+    rbtree_node* left_;
+    rbtree_node* parent_;
+    char mColor;
+  };
+
+  using node = rbtree_node;
+
+ protected:
+  struct rbtree_iterator {
+   public:
+    rbtree_iterator() noexcept;
+    rbtree_iterator(node* node) noexcept;
+    rbtree_iterator(const rbtree_iterator& other) : current_(other.current_) {};
+    reference operator*() const;
+    pointer operator->() const;
+    rbtree_iterator& operator++();
+    rbtree_iterator operator++(int);
+    rbtree_iterator& operator--();
+    rbtree_iterator operator--(int);
+    bool operator==(const rbtree_iterator& lhs);
+    rbtree_iterator& operator=(const rbtree_iterator& other) {
+      if (this != &other) {
+        current_ = other.current_;
+      }
+      return *this;
+    }
+    bool operator!=(const rbtree_iterator& lhs);
+
+   private:
+    friend class rbtree;
+    node* current_;
+  };
+
+  struct rbtree_const_iterator {
+   public:
+    rbtree_const_iterator() noexcept : rbtree_iterator() {};
+    const_reference operator*() const { return rbtree_iterator::operator*(); };
+  };
+
+  using iterator = rbtree_iterator;
+  using const_iterator = rbtree_const_iterator;
+
+ protected:
+  rbtree() noexcept;
+  rbtree(node* node);
+  rbtree(const std::initializer_list<value_type>& items);
+  rbtree(const rbtree& other) noexcept;
+  rbtree(rbtree&& other) noexcept;
+  void swap(rbtree<K, T>&& other);
+  void swap(rbtree<K, T>& other);
+  virtual ~rbtree();
+
+  rbtree& operator=(const rbtree& other) noexcept;
+  rbtree& operator=(rbtree&& other) noexcept;
+
+  iterator begin() noexcept;
+  iterator end() noexcept;
+
+  std::pair<iterator, bool> insert(const value_type& value) noexcept;
+  std::pair<iterator, bool> insert(const K& key, const T& obj);
+  std::pair<iterator, bool> insert_or_assign(const K& key, const T& obj);
+  void erase(iterator pos);
+  void erase(const value_type& value);
+  void merge(rbtree& other);
+  T& at(const K& key);
+  T& operator[](const K& key);
+  bool contains(const K& key);
+  size_type size();
+  void print_tree() { print_tree(m_root, 0); }
+  void clear();
+
+  node* minimum(node* n);
```

Comment:
Методы лучше называть глаголом. Например, тут лучше будет так

```suggestion
  node* get_min(node* n);
```
или
```suggestion
  node* find_min(node* n);
```